Player hotfixes and improvements

ferferga opened the pull request:
**Fixes**

* Playlist not loading when item doesn't have an image (this also fixes library when using ``ListView`` mode). Fixes #1429 and #1431 
* Playlist doesn't disappear anymore when editing/removing items from the queue.
* Doubled context menu. Fixes #1312 
* Playback buttons are always accesible without scrolling on mobile layout. Fixes #1195 [(this specific comment)](https://github.com/jellyfin/jellyfin-web/issues/1195#issuecomment-629692880)
* Fix some overlap issues.
* Fix jumpy elements when volume slider was not visible.
* Show all artists for a song. Each one of them is clickable.
* Remove space in miniplayer when item doesn't have image. *(Screenshots from mobile, but this was also happening in Desktop and TV as well. All layouts fixed)*
**Before**
![image](https://user-images.githubusercontent.com/10274099/85009855-6acc3a00-b15f-11ea-9918-63edfeb9a4f0.png)
**After**
![image](https://user-images.githubusercontent.com/10274099/85009762-440e0380-b15f-11ea-9271-5ed4cead14a0.png)

* Fix wrong contextmenu (from the miniplayer) appearing when clicking on an specific place in the right bottom corner
* Replaced context menu action ``ViewArtist`` for ``ViewAlbumArtist`` as specified in the code's comments:
```
// Show Album Artist by default, as a song can have multiple artists, which specific one would this option refer to?
// Although some albums can have multiple artists, it's not as common as songs.
```
* *Some other stuff I'm missing and I don't remember right now :P*

**Improvements**

* Better mobile layout.
* Add music-like shuffle mode. **Not tested in SyncPlay or remote sessions, but it will probably work out of the box**.
  * When enabling shuffle, old playlist will be saved in an array, so it will be used again when shuffle mode is disabled.
* Playlist is always shown in desktop layout. Fixes #1348
* Use our "native" context menu instead of a custom one. Fixes #1195 
* Revamped mobile miniplayer.
* Back button works like it does in most music players **(only audio affected, video is unaffected)**:
  * When more than 5 seconds of the song has been playing, player will start the current track from the beginning.
  * On double click after 5 seconds of the song, player will return to previous track.
  * If no previous item, no matter current song position, track will start from the beginning.

*I didn't build an apk with my branch, but this probably will need to be backported to Android and Expo, so OS media controls work. CC @jellyfin/client*

### Screenshots

**Revamped mobile miniplayer**
![image](https://user-images.githubusercontent.com/10274099/85000347-90067b80-b152-11ea-8b70-64b46644e54a.png)

*Context menu was replaced by the next track button*

**Full screen mobile miniplayer**
![image](https://user-images.githubusercontent.com/10274099/85000526-d0fe9000-b152-11ea-9e28-967394cef5a1.png)

*Removed rewinds and add repeat and shuffle commands*

**Desktop miniplayer**
![image](https://user-images.githubusercontent.com/10274099/85000749-20dd5700-b153-11ea-9202-6991e606d8cd.png)

*Playlist shown by default*

**Shuffle**
![shuffle](https://user-images.githubusercontent.com/10274099/85000921-55511300-b153-11ea-8ddc-79742455bf7f.gif)

*Old playlist is reused after disabling shuffle*

**Back button behaviour**
![back-button](https://user-images.githubusercontent.com/10274099/85002097-ca711800-b154-11ea-9657-e156826dcfe2.gif)

*Behaviour of the back button as explained above. Last press to back button is a double click*

**Related Issues and PRs**
Closes #923 
Closes #924
Completes and add fixes #1056
Closes #1491
Closes #1456

ferferga commented:
Although this PR seems to be more feature-wise than fixing-wise, most of the improvements here actually make the new layout usable, so I think this is fitting enough for 10.6. Feel free to disagree with me by commenting in this PR.

ferferga commented on src/components/listview/listview.js:
```diff
@@ -270,52 +270,54 @@ define(['itemHelper', 'mediaInfo', 'indicators', 'connectionManager', 'layoutMan
+                    if (imgUrl) {
+                        html += '<div data-action="' + imageAction + '" class="' + imageClass + ' lazy" data-src="' + imgUrl + '" ' + blurhashAttrib + ' item-icon>';
+                    } else {
+                        html += '<div class="' + imageClass + '">';
+                    }
 
-                var indicatorsHtml = '';
-                indicatorsHtml += indicators.getPlayedIndicatorHtml(item);
+                    var indicatorsHtml = '';
+                    indicatorsHtml += indicators.getPlayedIndicatorHtml(item);
 
-                if (indicatorsHtml) {
-                    html += '<div class="indicators listItemIndicators">' + indicatorsHtml + '</div>';
-                }
+                    if (indicatorsHtml) {
+                        html += '<div class="indicators listItemIndicators">' + indicatorsHtml + '</div>';
+                    }
 
-                if (playOnImageClick) {
-                    html += '<button is="paper-icon-button-light" class="listItemImageButton itemAction" data-action="resume"><span class="material-icons listItemImageButton-icon play_arrow"></span></button>';
-                }
+                    if (playOnImageClick) {
+                        html += '<button is="paper-icon-button-light" class="listItemImageButton itemAction" data-action="resume"><span class="material-icons listItemImageButton-icon play_arrow"></span></button>';
+                    }
 
-                var progressHtml = indicators.getProgressBarHtml(item, {
-                    containerClass: 'listItemProgressBar'
-                });
+                    var progressHtml = indicators.getProgressBarHtml(item, {
+                        containerClass: 'listItemProgressBar'
+                    });
 
-                if (progressHtml) {
-                    html += progressHtml;
+                    if (progressHtml) {
+                        html += progressHtml;
+                    }
+                    html += '</div>';
                 }
-                html += '</div>';
```
With this, items that don't have images won't display anything. Example (from ``remotecontrol.js`` playlist):

![image](https://user-images.githubusercontent.com/10274099/85007059-48382200-b15b-11ea-9048-0d5a15fa6bf3.png)

Shall we use placeholder random backgrounds and logos? (As it's done in ``cardbuilder.js``)

MrTimscampi replied:
I'd remove the image for every item in that list, especially as there is already the album art near there.

But it's open for discussion, of course :)

dkanada replied:
I think we should always show the album art, because often queues will have lots of different material. It would also look weird to have some items with an image and some without. We should really just use the placeholder image for the lists.

I **am** in agreement that the image removal is fine on the playback bar though, since it only has a single image.

JustAMan replied:
I support @dkanada here, hiding an image of _some entries_ in a list would make list look like a patchwork, let's use placeholders there. Hiding empty image in a single-item view is fine for me.

ferferga commented on src/components/playback/playbackmanager.js:
```diff
@@ -3811,7 +3819,7 @@ define(['events', 'datetime', 'appSettings', 'itemHelper', 'pluginManager', 'pla
         });
     };
 
-    PlaybackManager.prototype.shuffle = function (shuffleItem, player, queryOptions) {
```
Did a bit of cleaning as well in ``playbackManager``. This was removed because it's unused in the function.

ferferga commented on src/strings/es.json:
```diff
@@ -1560,5 +1560,6 @@
     "ShowMore": "Mostrar más",
-    "ShowLess": "Mostrar menos"
+    "ShowLess": "Mostrar menos",
+    "ViewAlbumArtist": "Ver artista del álbum"
```
@dmitrylyzo Good catch!

ferferga commented:
My last commit improves the overall look of all the interface when resizing the window using desktop layout as miniplayer and mobile.

I want to keep improving the layout and make one that works well enough on tablets, but definitely not for 10.6

hamburglar2160 commented:
I haven't been around a cast device to capture any logs yet but #1346 falls in the 'bug fixing' and #1349 in the 'finishing up revamp/improvements scope of this PR.

KristupasSavickas commented:
>Show all artists for a song. Each one of them is clickable.

This is huge, thanks!

dkanada commented:
@ferferga sometimes I add ten or fifteen lines to a pull request but this seems way too large to fit into a single pull request. There are also **tons** of style changes that could easily wait until the next minor release.

ferferga commented:
@dkanada Most of the style changes you see were already around our code since #1056 was merged. I just moved them outside the ``@media`` rule and applied them only with the ``mobilePlayer`` class (which is applied on ``remotecontrol.js`` when using mobile layout, making everything much more predictable). I only added the new shuffle button bits, the transparency class to solve [this](https://github.com/jellyfin/jellyfin-web/issues/1195#issuecomment-629692880) and fixed the margins in the miniplayer for songs without images and with images. Those changes are necessary so the text doesn't overflow the buttons with the newly shuffle one. 

I know shuffle could've gone into 10.7 instead, but adding the repeat button was something that was mentioned in the original PR as necessary to be accomplished to finish that up (It's coming too late, though). I did that, but it looked really awkward without no button on the right side, so a shuffle mode was necessary. Most of the things that needed fixes were either a few-line changes or doesn't make sense with each other apart.

I tested with all the possible layouts and the default viewports that Chrome & Firefox allows you to use, and I see no issues so far. I'm cool with this going for 10.6.1 but, **sincerely, I don't see the reason to release a half-baked player (+ the blog post announcement that will include it I guess) to deliver the full experience in a minor update later when we can do it directly for 10.6.0. We still have a week ahead for reviewing this if we're aiming at the 26th due date**

ferferga commented:
@dmitrylyzo Thank you very much for testing:

I'm going to do all your requested changes and test them on Monday, after finishing my exams.

Yes, that's the only caveat that Shuffle mode has at the moment. Before opening this PR, I did some tests with a draft implementation of shuffle mode and everything seemed that was going to be really complicated and it wouldn't take too much time and not be ready for 10.6. Joining all the pieces together turned out to be easier than expected though, so I moved everything to the remote branch and pushed it as is. I can start working on it on Monday. However, I don't think I should do that in this PR: **As pointed by @dkanada and @JustAMan, this PR is quite big already** and I rather have this reviewed and merged for 10.6 (or reviewers will hate me after adding more features here, instead of the requested changes 😁) and **leave that alone in another PR**. Although shuffle isn't 1-1 feature-wise compared with other apps, I think it's in an state good enough to be released in stable imho

About the last point, I'm not sure what you mean: In Spotify and other music apps, if I shuffle, my current queue shuffles completely. If I disable shuffle, the original queue goes back. Shuffling again generates a completely new queue. This is how it works currently. Can you give me examples with other apps?

dmitrylyzo commented:
> Can you give me examples with other apps?

_I can speak for PC apps_
I used Amarok for a long time until it was deprecated in Kubuntu (due to Qt4). It has random play ~ shuffle on each playlist repeat (as I suggested). Now I use Cantata+MPD with same logic.

In addition to my suggestion, you can shuffle the entire list, but move the current item to the beginning.

ferferga commented:
@dmitrylyzo I was talking about the last point. The first one it's perfectly clear :).

What do you mean with "It's applied once exactly?"

dmitrylyzo commented:
> What do you mean with "It's applied once exactly?"

You enable `Repeat`.
You enable `Shuffle`.
It shuffles playlist.
It plays entire list.
It starts from the beginning, but queue will be the same again and again. (_"Shuffle" is applied once_)

ferferga commented:
@dmitrylyzo Now I understand

* Your first point applies to all players I used as well, so no discussions here

* This is the first time I heard about this behaviour. Based on the players I tested:

    * *On Spotify, Google Play Music, Deezer, Tidal and Apple Music*: Shuffling does show you the resulting playlist, so you can see what's coming next. When it finishes, if no repeat, music will stop. If repeat, the playlist will start again from the beginning (the track you were listening when you hit shuffle) without further shuffling. Aside from *the track you were listening when you hit shuffle*, **this is the shuffling concept I developed for Jellyfin**. Something different than this is a bug.

    * I rechecked on **VLC** after a long time without using it (looking at you, streaming services): Shuffling doesn't result in a different playlist, visually. VLC simply takes a random song from the list. When all the songs of the playlist have been played, **even without repeat enabled** you will end up relistening to the whole playlist over and over, as the player doesn't take in account what we have listened: **it just picks items randomly**. Enabling *Repeat All* mode doesn't make sense in this context, just *Repeat One* changes how the player behaves (playing the current song in a loop).

* **Phonograph/Gelli**: Same as streaming services.

Although I'm used to the "new" shuffling methods of the streaming services, using VLC brought me good memories of the old times when I started listening to music: it was Windows Media Player with a (back then) really modern Windows XP. Doing these tests remembered me how it worked, in exactly the same way, and, to be honest, I miss it a lot. As a side note, I think Spotify (back when it was invite only) did it in the same way.

Given these results, I think you'll immediately agree with me that we should leave this behaviour intact for now, people will probably be used to what they have in other platforms. However, I'm 100% open (and plan to) add an option in settings, so users can switch between shuffling methods. But, as in my other message, that's for another PR :).

Is that OK in your opinion?

Btw, it's really cool to see how something as small as shuffle can be understood so differently between apps and people :P

dmitrylyzo commented:
> Given these results, I think you'll immediately agree with me that we should leave this behaviour intact for now, people will probably be used to what they have in other platforms. However, I'm 100% open (and plan to) add an option in settings, so users can switch between shuffling methods. But, as in my other message, that's for another PR :).
> 
> Is that OK in your opinion?

I agree with option.
I am just used to turning on music and forget about re-shuffling. :stuck_out_tongue:

ferferga commented on src/components/nowPlayingBar/nowPlayingBar.js:
```diff
@@ -545,21 +586,27 @@ define(['require', 'datetime', 'itemHelper', 'events', 'browser', 'imageLoader',
+                        // We remove the previous event listener by replacing the item in each update event
+                        let contextButtonClone = contextButton.cloneNode(true);
+                        contextButton.parentNode.replaceChild(contextButtonClone, contextButton);
+                        contextButton = document.querySelector('.nowPlayingBar').querySelector('.btnToggleContextMenu');
```
@dmitrylyzo Here seems to work, but not in ``remotecontrol.js``. I guess it's because we reconstruct the whole DOM in the miniplayer. I'm going to leave them as I did for the sake of consistency.

I plan on revisit the players later, so this might be changed in the future.

ferferga commented:
## Changes in latest commits

* Styling cleanup
* Items in queue without images will display a placeholder.
* Cards placeholders now display **Audio** items with *audiotrack* icon
* Current playing items will be kept first when shuffled. Previous item in queue will also be shuffled (this might get improvements later)
* Fixed Chromecast and remote players reporting ``undefined``. Fixes #1346 
* Player can now sent shuffle actions to Chromecast and remote players now support shuffle mode.

*This avoids a crash in the player when casting. Chromecast still needs to have this implemented in their own code, something which I'm going to do soon. What I have done is allow playbackManager to **send** the "signal" to the remote players, but remote players also need to have this implemented*. **This means that shuffling when casting to Chromecast doesn't do anything (but doesn't crash as well)**

MrTimscampi commented on src/components/remotecontrol/remotecontrol.js:
```diff
@@ -650,12 +708,40 @@ define(['browser', 'datetime', 'backdrop', 'libraryBrowser', 'listView', 'imageL
+
+            context.querySelector('.btnPreviousTrack').addEventListener('click', function (e) {
+                if (currentPlayer) {
+                    if (currentPlayer.id === 'htmlaudioplayer' && (currentPlayer._currentTime >= 5 || !playbackManager.previousTrack(currentPlayer))) {
```
```suggestion
                    if (currentPlayer.id === 'htmlaudioplayer' && (currentPlayer._currentTime >== 5 || !playbackManager.previousTrack(currentPlayer))) {
```

ferferga replied:
@MrTimscampi This gives ``unexpected token error``

MrTimscampi commented on src/components/itemContextMenu.js:
```diff
@@ -288,10 +288,11 @@ define(['apphost', 'globalize', 'connectionManager', 'itemHelper', 'appRouter',
+        if (options.openArtist !== false && item.AlbumArtists && item.AlbumArtists.length) {
             commands.push({
-                name: globalize.translate('ViewArtist'),
+                name: globalize.translate('ViewAlbumArtist'),
```
I do think that it should send to the album artist, though.

ferferga replied:
@MrTimscampi This is something we should change in Weblate instead, right? Or you mean replacing the key?

I just reused the ViewArtist one (renamed it so it gets retranslated). Or you mean creating an entirely new key? In that case, we should remove the old one.

dkanada commented on src/components/playback/playbackmanager.js:
```diff
@@ -4000,6 +4030,9 @@ define(['events', 'datetime', 'appSettings', 'itemHelper', 'pluginManager', 'pla
             case 'SetRepeatMode':
                 this.setRepeatMode(cmd.Arguments.RepeatMode, player);
                 break;
+            case 'SetQueueShuffleMode':
+                this.setQueueShuffleMode(cmd.Arguments.ShuffleMode, player);
+                break;
```
Why not match `SetRepeatMode` and label this `SetShuffleMode` to keep it more generic for now? This is an excellent change though, I'm always a fan of better command support.

ferferga replied:
@dkanada That's how it was called initially. However, didn't want to make a confusion between the "normal" shuffle (the one you run in the library browser, item details, context menu, etc) and the queue one. @jelyfin/web thoughts?

MrTimscampi replied:
I think all these should eventually be merged, so reusing the same name shouldn't be an issue.

dkanada replied:
Were you going to add this in a future commit?

ferferga replied:
@dkanada Is already done in b7714ee

ferferga commented on src/components/itemContextMenu.js:
```diff
@@ -458,7 +459,7 @@ define(['apphost', 'globalize', 'connectionManager', 'itemHelper', 'appRouter',
                     break;
                 case 'artist':
-                    appRouter.showItem(item.ArtistItems[0].Id, item.ServerId);
+                    appRouter.showItem(item.AlbumArtists[0].Id, item.ServerId);
```
@MrTimscampi We already check that in the function above, where we push options to the array that builds the context menu. If no Album Artist, no option will be displayed, so no way this would be executed.

ferferga commented on src/components/nowPlayingBar/nowPlayingBar.js:
```diff
@@ -146,13 +157,38 @@ define(['require', 'datetime', 'itemHelper', 'events', 'browser', 'imageLoader',
+                        return;
+                    }
+                    playbackManager.seekPercent(0, currentPlayer);
+                    // This is done automatically by playbackManager, however, setting this here gives instant visual feedback
```
@MrTimscampi The default for the dblclick JS event. We can set our own, but IMO it's fine and on par with what you would expect in all other music apps.

For context, the delay it has is the same as the one you have when double tapping on any mobile browser and it zooms.

Check https://developer.mozilla.org/en-US/docs/Web/API/Element/dblclick_event example also.

MrTimscampi replied:
Hm, it's nearly instantaneous here. It should be mostly fine if we let playbackManager handle the position instead of forcing it.

ferferga replied:
@MrTimscampi My bad, I thought you meant the back behaviour. Here is a video in Chrome (maybe in FF it's faster and you didn't notice it, I can say in Chrome it is). Maybe I'm too nitpicky, but I'm able to tell the difference easily

Both videos hit back at the 1:40 timestamp. Bear in mind with the low framerates of gifs, it's not the best way to show this here, to be honest.

**With "manual" change:**
![after](https://user-images.githubusercontent.com/10274099/85418971-ca688200-b571-11ea-802a-d1f69700c2e5.gif)

**With ``playbackmanager``**:
![before](https://user-images.githubusercontent.com/10274099/85419272-2af7bf00-b572-11ea-8f97-8ef4297a6461.gif)

MrTimscampi replied:
Hm, I think the difference is negligible and it's better to not have stuff interacting with something that's controlled in a central component without going through that component, imo.

It's not blocking, but I think it's better design-wise to let the playbackManager handle the seekbar :p

ferferga replied:
@MrTimscampi I'll add a TODO so we can investigate later the delay that playbackManager has in the seek function. I'm full for this because makes the difference between a webpage and the "sensation" that we're in a native app, which I think Android users will appreciate.

ferferga commented on src/components/nowPlayingBar/nowPlayingBar.js:
```diff
@@ -146,13 +157,38 @@ define(['require', 'datetime', 'itemHelper', 'events', 'browser', 'imageLoader',
 
+        elem.querySelector('.btnShuffle').addEventListener('click', function () {
+            if (currentPlayer) {
+                if (playbackManager.getPlaylistShuffleMode(currentPlayer) === 'Sorted') {
```
@MrTimscampi FYI this was my initial implementation, in order to keep things simple :). Changed it to using values to keep consistency (regarding Repeat modes) and also because I thought that maybe external third party players that use our playbackManager or some crazy functionality we add in the future could simply keep shuffling without switching modes.

Didn't thought about re-adding it as a helper function and I think it's excellent because its simple enough and gives full control.

Is there any problem if I add it and simply leave it unused? Or shall I change all my references in remotecontrol and nowplayingbar to it?

MrTimscampi replied:
Ideally, nothing should be unused ;)

If it's not much work, feel free to add it and change references. Otherwise, simply adding a TODO will do for now :)

ferferga commented on src/components/playback/playqueuemanager.js:
```diff
@@ -56,6 +58,36 @@ define([], function () {
+            this._playlist[i] = this._playlist[j];
+            this._playlist[j] = temp;
+        }
+        this._playlist.unshift(currentPlaylistItem);
```
@dmitrylyzo Yes, in the splice. Afterwards, with the continue, it goes to the following item.

After adding this I did a search multiple times and I couldn't find duplicate items. You're saying this because you actually had this happening?

dmitrylyzo replied:
> You're saying this because you actually had this happening?

Yes.
I placed breakpoints in "remove current" block and before `unshift`.
It didn't stop in "remove current" block.
Before `unhisht`, I still have current item.

Fun fact, when current item is shuffled at the end of queue, the latter is virtually looped - player switches to "first" current item instead of shuffled last.

ferferga replied:
@dmitrylyzo Check how I managed this in my next commits. I removed it before shuffling and then added it afterwards.

dmitrylyzo replied:
It sounds promising.

ferferga commented on src/components/playback/playqueuemanager.js:
```diff
@@ -176,21 +208,39 @@ define([], function () {
+        if (value === 'RepeatOne' || value === 'RepeatAll' || value === 'RepeatNone') {
+            this._repeatMode = value;
+        } else {
+            throw new TypeError('invalid value provided for setRepeatMode');
```
@MrTimscampi Yeah, with the current code is impossible as well. I thought about adding this here basically in case someone wants to touch something in the remote players and don't have a clue why shuffling doesn't work :)

ferferga commented on src/components/remotecontrol/remotecontrol.css:
```diff
@@ -157,42 +157,106 @@
+    border-radius: 0;
+}
+
+.layout-mobile .playlistSectionButton .volumecontrol {
+    margin: 0;
+    padding-right: 0;
+    border-radius: 0;
+}
```
@dmitrylyzo This is the original styling that was applied in #1056, I just applied only to mobile layout. For the padding, done in my next commits.

ferferga commented on src/components/cardbuilder/card.css:
```diff
@@ -167,8 +167,9 @@ button::-moz-focus-inner {
+}
 
-    /* This is only needed for scalable cards */
+.cardImageContainer.cardScalable {
```
@dmitrylyzo Good catch!

ferferga commented on src/components/cardbuilder/cardBuilder.js:
```diff
@@ -1534,8 +1534,11 @@ import 'programStyles';
                 case 'MusicAlbum':
                     return '<span class="cardImageIcon material-icons album"></span>';
                 case 'MusicArtist':
+                    return '<span class="cardImageIcon material-icons person"></span>';
```
@dmitrylyzo Ooops! Didn't know this about the switch syntax.

dmitrylyzo commented on src/components/remotecontrol/remotecontrol.js:
```diff
@@ -316,25 +354,25 @@ define(['browser', 'datetime', 'backdrop', 'libraryBrowser', 'listView', 'imageL
-                toggleRepeatButton.classList.remove('repeatButton-active');
+            let toggleRepeatButtons = context.querySelectorAll('.repeatToggleButton');
+
+            for (let toggleRepeatButton of toggleRepeatButtons) {
+                if ('RepeatAll' == repeatMode) {
+                    toggleRepeatButton.innerHTML = "<span class='material-icons repeat'></span>";
+                    toggleRepeatButton.classList.add('repeatButton-active');
+                } else if ('RepeatOne' == repeatMode) {
+                    toggleRepeatButton.innerHTML = "<span class='material-icons repeat_one'></span>";
+                    toggleRepeatButton.classList.add('repeatButton-active');
+                } else {
+                    toggleRepeatButton.innerHTML = "<span class='material-icons repeat'></span>";
+                    toggleRepeatButton.classList.remove('repeatButton-active');
+                }
```
_17 hours later_
Oh, wait. Won't work with `classList` because of add/remove.

dmitrylyzo commented on src/components/nowPlayingBar/nowPlayingBar.js:
```diff
@@ -501,21 +534,30 @@ define(['require', 'datetime', 'itemHelper', 'events', 'browser', 'imageLoader',
-            return `<div ${cssClass}>${nowPlayingText}</div>`;
-
-        }).join('');
+            if (textLines[0].text) {
+                if (textLines[0].text) {
```
overdid :p
also you could leave creation order: `itemText`, `secondaryText`

ferferga replied:
@dmitrylyzo Lol, not sure what I was thinking here :P

dmitrylyzo commented on src/components/remotecontrol/remotecontrol.js:
```diff
@@ -118,19 +120,30 @@ define(['browser', 'datetime', 'backdrop', 'libraryBrowser', 'listView', 'imageL
+                        // to normal item.Artists item.
+                        // TODO: Normalise fields returned by all the players
+                        for (const artist of item.Artists) {
+                            artistsSeries += `<a>${artist}</a>`;
```
not `emby-linkbutton`, just `a`?

ferferga replied:
@dmitrylyzo Yes, because Chromecast doesn't return ``ArtistItems`` for some reason. No way I could generate a link without the id of the artist.

dmitrylyzo commented on src/components/remotecontrol/remotecontrol.js:
```diff
@@ -316,25 +353,37 @@ define(['browser', 'datetime', 'backdrop', 'libraryBrowser', 'listView', 'imageL
+            let toggleRepeatButtons = context.querySelectorAll('.repeatToggleButton');
+            const cssClass = 'repeatButton-active';
+            let innHtml = '<span class="material-icons repeat"></span>';
+            let repeatOn = undefined;
```
Why didn't you use boolean values?

ferferga replied:
@dmitrylyzo Same as above, not sure what I was thinking here :P

dmitrylyzo commented:
I got `throw new TypeError('invalid value for repeatMode');` on stop.

ferferga commented:
@dmitrylyzo Hmm true, because at stop it becomes ``undefined``. Shall we keep the switch or better revert back to the ``if`` structure? Not sure what's the best approach now.

dmitrylyzo commented on src/components/playback/playqueuemanager.js:
```diff
@@ -127,6 +155,12 @@ define([], function () {
         var currentPlaylistItemId = this.getCurrentPlaylistItemId();
         var isCurrentIndex = playlistItemIds.indexOf(currentPlaylistItemId) !== -1;
 
+        if (this._sortedPlaylist.length <= playlistItemIds.length) {
+            this._sortedPlaylist = this._sortedPlaylist.splice(0).filter(function (item) {
+                return playlistItemIds.indexOf(item.PlaylistItemId) === -1;
+            });
+        }
```
`playlistItemIds` is items to remove, so `this._sortedPlaylist.length <= playlistItemIds.length` doesn't work.
Also I see no reason to use `splice`.
```suggestion
        this._sortedPlaylist = this._sortedPlaylist.filter(function (item) {
            return !playlistItemIds.includes(item.PlaylistItemId);
        });
```

ferferga replied:
@dmitrylyzo The condition is wrong, correct. I added it after having it working and, although I retested everything before pushing, for some reason it seems that I missed this.

* ``splice`` is required, otherwise the resulting array will be empty. In this function, the playlist variable (returned from ``getCurrentPlaylist`` is already spliced, that's why not used later. Just tested and it doesn't work without splice.

* Also, we must check against -1 index, no clue why but your return doesn't work as well.

What I did here is replicate with the sorted playlist array what it's done in the current playlist. It's not the ideal solution but we're too late in the feature freeze and this is not yet migrated to ES6, so I think it makes sense to leave all the refactoring for later (if nobody comes first, I want to migrate the playback components to ES6, already got some appreciation to them :P) and stay in the safe side, even in the little details. Didn't inspect the objects deeply as well to improve this, will do on ES6 migration.

dmitrylyzo replied:
>     * `splice` is required, otherwise the resulting array will be empty.

Hmm, but it works without it. Moreover, `filter` returns a filtered copy.

ferferga replied:
I see what is the problem now 👍

dmitrylyzo commented on src/components/remotecontrol/remotecontrol.js:
```diff
@@ -316,25 +353,39 @@ define(['browser', 'datetime', 'backdrop', 'libraryBrowser', 'listView', 'imageL
+            }
+
+            for (const toggleRepeatButton of toggleRepeatButtons) {
+                if (!repeatOn) {
+                    toggleRepeatButton.classList.remove(cssClass);
+                } else {
+                    toggleRepeatButton.classList.add(cssClass);
+                }
```
```suggestion
                toggleRepeatButton.classList.toggle(cssClass, repeatOn);
```
In fact, this can be applied in many cases that we have.

ferferga replied:
@dmitrylyzo Didn't know about ``toggle``! This is a life saver! 😱 Thank you!

Will apply it only to the new toggles I added/edited in this PR. Will keep all the ones that were previously in this file intact, don't want to make the diffs any larger. Will take a note of this for the ES6 migration.

dmitrylyzo commented on src/components/remotecontrol/remotecontrol.js:
```diff
@@ -470,6 +523,28 @@ define(['browser', 'datetime', 'backdrop', 'libraryBrowser', 'listView', 'imageL
         }
 
+        function onShuffleQueueModeChange() {
+            let shuffleMode = playbackManager.getQueueShuffleMode(this);
```
Maybe `playbackManager.getQueueShuffleMode` shouldn't return an invalid value? And we can rely on it.

ferferga replied:
@dmitrylyzo I think I know now what was causing us the issue in the first place. No need to use ``undefined`` in the switch anymore, repeat values from playbackmanager will always be ``RepeatNone``, ``RepeatAll`` and ``RepeatOne``. Same for shuffle. Let me test that and hopefully this get solved in next commit.

dmitrylyzo commented on src/nowplaying.html:
```diff
@@ -163,20 +177,17 @@ <h2 class="nowPlayingPageTitle"></h2>
         </div>
         <div class="playlistSection">
             <div class="playlistSectionButton flex align-items-center justify-content-center">
```
For TV. Keep focus inside container with horizontal keyboard navigation.
```suggestion
            <div class="playlistSectionButton flex align-items-center justify-content-center focuscontainer-x">
```

ferferga commented:
@dmitrylyzo Hmm, TV layout doesn't have nowplayingbar, only remotecontrol component 😨. At least, never saw it. You switched while running?

No idea why that happens there

ferferga commented on src/components/remotecontrol/remotecontrol.js:
```diff
@@ -465,9 +511,29 @@ define(['browser', 'datetime', 'backdrop', 'libraryBrowser', 'listView', 'imageL
+            for (let shuffleButton of shuffleButtons) {
+                switch (shuffleMode) {
+                    case 'Sorted':
+                        shuffleButton.classList.toggle(cssClass, false);
```
@dmitrylyzo It does the work right though

ferferga commented on src/components/remotecontrol/remotecontrol.js:
```diff
@@ -252,14 +252,16 @@ define(['browser', 'datetime', 'backdrop', 'libraryBrowser', 'listView', 'imageL
                 case 'RepeatOne':
                     playbackManager.setRepeatMode('RepeatNone');
+                    break;
+                default:
```
@dmitrylyzo Given that we receive inconsistent responses from playbackManager, I thought it was a better idea to have all the cases handled, at least until we have some consistency in playbackManager. However, I've been playing with this a little and it seems fine, so I'm going to remove them as you suggest.

ferferga commented on src/components/remotecontrol/remotecontrol.js:
```diff
@@ -362,24 +404,28 @@ define(['browser', 'datetime', 'backdrop', 'libraryBrowser', 'listView', 'imageL
+                var nowPlayingVolumeSliderContainer = context.querySelector('.nowPlayingVolumeSliderContainer');
+
+                if (nowPlayingVolumeSlider) {
+                    if (showVolumeSlider) {
+                        nowPlayingVolumeSliderContainer.classList.remove('hide');
+                    } else {
+                        nowPlayingVolumeSliderContainer.classList.add('hide');
+                    }
```
@dmitrylyzo Little change, so did it anyway. *Why not?* :P

dmitrylyzo commented on src/components/remotecontrol/remotecontrol.js:
```diff
@@ -465,9 +503,28 @@ define(['browser', 'datetime', 'backdrop', 'libraryBrowser', 'listView', 'imageL
+            for (let shuffleButton of shuffleButtons) {
+                switch (shuffleMode) {
+                    case 'Shuffle':
+                        shuffleButton.classList.toggle(cssClass, true);
```
```suggestion
                        shuffleButton.classList.add(cssClass);
```

dmitrylyzo commented on src/components/remotecontrol/remotecontrol.js:
```diff
@@ -465,9 +503,28 @@ define(['browser', 'datetime', 'backdrop', 'libraryBrowser', 'listView', 'imageL
+                        break;
+                    case 'Sorted':
+                    default:
+                        shuffleButton.classList.toggle(cssClass, false);
```
```suggestion
                        shuffleButton.classList.remove(cssClass);
```

ferferga commented:
@dmitrylyzo I think latest commits already make the playlist reload only when necessary. Still, this component needs lots of works, so the whole playlist is updated only when necessary, but not for this PR imo.

I also fixed the focus issue on TV (also applies to the rest of layouts, so in the future we can add some kind of indication for the active item on desktop layout)

ferferga commented on src/components/nowPlayingBar/nowPlayingBar.js:
```diff
@@ -159,7 +159,7 @@ define(['require', 'datetime', 'itemHelper', 'events', 'browser', 'imageLoader',
         elem.querySelector('.previousTrackButton').addEventListener('click', function (e) {
             if (currentPlayer) {
-                if (currentPlayer.id === 'htmlaudioplayer' && (currentPlayer._currentTime >= 5 || !playbackManager.previousTrack(currentPlayer))) {
+                if (lastPlayerState.NowPlayingItem.MediaType === 'Audio' && (currentPlayer._currentTime >= 5 || !playbackManager.previousTrack(currentPlayer))) {
```
@dmitrylyzo I have a future plan of lending all the music and audio playback to **howler**, so we can switch between HTML Audio and WebAudio + a few other goodies like gapless playback (*The library to use ofc needs discussion in Matrix and so on of course*). With this check, we're **explicitly checking for HTML Audio**, so other audio plugins/"players" like WebAudio, howler, etc are excluded. However, if we explicitly check for the item type, we're 100% sure that what we're playing is Audio, no matter what we use to play it.

At first, I thought about making the change later, once we roll those changes to the music players. However, no harm in doing this now that we're polishing this PR :P

ferferga commented on src/components/remotecontrol/remotecontrol.js:
```diff
@@ -163,11 +176,29 @@ define(['browser', 'datetime', 'backdrop', 'libraryBrowser', 'listView', 'imageL
-            console.debug('updateNowPlayingInfo');
+            let contextButton = context.querySelector('.btnToggleContextMenu');
+            // We remove the previous event listener by replacing the item in each update event
+            let contextButtonClone = contextButton.cloneNode(true);
+            contextButton.parentNode.replaceChild(contextButtonClone, contextButton);
+            contextButton = context.querySelector('.btnToggleContextMenu');
```
@dmitrylyzo As I said [here](https://github.com/jellyfin/jellyfin-web/pull/1430#discussion_r443620102), it doesn't seem to work without the replace child, so I'm going to commit your suggestion straightaway, without removing it.

dmitrylyzo replied:
Looks like `contextButton` exists in the template.
As a possible solution, you could extract options that passed into `itemContextMenu.show` (`itemContextMenuOptions`) and update them here. `contextButton.addEventListener` goes  to `bindEvents`.
No rush - you can do it in the next PR.

ferferga commented:
@dmitrylyzo *For some reason I can't reply you to the review comment, so I do it here*:

Okay, I'll note this for a future PR.

Last 3 commits adds the following:

* Stop playback option in context menu (only mobile layout)
* Clear queue option in context menu
![image](https://user-images.githubusercontent.com/10274099/86252873-0d4ed900-bbb4-11ea-94d5-f8a35e0020f7.png)

**Borrar la cola**: Clear queue
**Detener la reproducción**: Stop playback
* Option parity between miniplayer and remotecontrol in context menu: *Before, miniplayer showed more options, like download, see media info, etc. Now, those are also in remotecontrol, as you can see*

CC @dmitrylyzo @MrTimscampi and @dkanada

dmitrylyzo commented on src/components/remotecontrol/remotecontrol.js:
```diff
@@ -163,11 +176,38 @@ define(['browser', 'datetime', 'backdrop', 'libraryBrowser', 'listView', 'imageL
+            if (autoFocusContextButton) {
+                contextButton.focus();
+            }
+            let stopPlayback = false || layoutManager.mobile;
```
`false` seems unnecessary.
Do you want a guaranteed boolean in case of `undefined`?
```suggestion
            const stopPlayback = !!layoutManager.mobile;
```

ferferga commented:
/azp run

dmitrylyzo commented on src/components/playback/playbackmanager.js:
```diff
@@ -2969,6 +2970,7 @@ define(['events', 'datetime', 'appSettings', 'itemHelper', 'pluginManager', 'pla
             } else {
                 self._playQueueManager.queue(items);
             }
+            events.trigger(player, 'playlistitemmove');
```
By meaning, `playlistitemadd` fits better. But it isn't handled.

ferferga replied:
@dmitrylyzo Didn't know that we had an event for that and didn't want to mess so much with them. I checked and added it to ``remotecontrol.js```

Good catch!
